gradle-plugin/src/main/kotlin/io/spine/protodata/gradle/plugin/Extension.kt

```diff
@@ -81,12 +83,64 @@ public class Extension(internal val project: Project): CodegenSettings {
     internal fun requestFile(forSourceSet: SourceSet): Provider<RegularFile> =
         requestFilesDirProperty.file(CodeGeneratorRequestFile.name(forSourceSet))
 
+    override val paths: Multimap<String, SourcePaths> = HashMultimap.create()
+
+    /**
+     * Obtains the source configured paths.
+     *
+     * If the deprecated [subDirs] and [targetBaseDir] are used, constructs [SourcePaths] instances
+     * from the present data for backward compatibility. However, in this compatibility mode,
+     * only Java and Kotlin source file sets can be constructed.
+     */
+    internal fun pathsOrCompat(sourceSet: SourceSet): Set<SourcePaths> {
+        if (!paths.isEmpty) {
+            return paths[sourceSet.name].toSet()
+        }
+        val srcRoots = sourceDirs(sourceSet).get()
+        val targetRoots = targetDirs(sourceSet).get()
+        return srcRoots.zip(targetRoots)
+            .map { (src, target) ->
+                val pathSuffix = src.asFile.toPath().name
+                val lang = if (pathSuffix == "kotlin") Kotlin else Java
+                val generatorName = if (pathSuffix.equals(lang.name, ignoreCase = true)) {
+                    ""
+                } else {
+                    pathSuffix
+                }
+                SourcePaths(
+                    src.asFile,
+                    target.asFile,
+                    lang,
+                    generatorName
+                )
+            }.toSet()
+    }
+
+    public companion object {
+
+        /**
+         * Default subdirectories expected by ProtoData under a generated source set.
+         *
+         * @see subDirs
+         */
+        public val defaultSubdirectories: List<String> = listOf(
+            "java",
```

Comment:
With all these words constantly repeating, it again strikes me on why we don't re-use the definitions, say, from either languages, or generators (or labels), that we know exist.

Reply:
Some here. This list will die off once the old `subdirs` API is removed.